fast-cdr: add version 2.3.0, fast-dds: add versions 2.14.4 and 3.2.1

nicosmd opened the pull request:
### Summary
Changes to recipe:  **fast-cdr/2.3.0** **fast-dds/2.14.4** **fast-dds/3.2.1**

#### Motivation
Add newest versions of Fast-DDS and Fast-CDR.

#### Details
Fast-DDS 3.2.1requires Fast-CDR Version 2.3.0 why it was added by the PR as well. 
Fast-DDS was splitted in an old 2.x folder and a new 3.x as some things in the conanfile changed and the test_package is significantly differnt.


---
- [x] Read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md)
- [x] Checked that this PR is not a duplicate: [list of PRs by recipe](https://github.com/conan-io/conan-center-index/discussions/24240)
- [x] Tested locally with at least one configuration using a recent version of Conan

jcar87 commented:
Hi @nicosmd - thanks for your PR.

> Fast-DDS was splitted in an old 2.x folder and a new 3.x as some things in the conanfile changed and the test_package is significantly differnt.

This makes the PR very difficult to review - we would appreciate if you could expand on "some things" and provide details so that this can be more easily reviewed, thanks!

nicosmd commented:
Hey @jcar87, sure! 

Let me summarize what I ment by "some things" in Fast-DDS:

- removed conan1 compatibility in new 3.X fast-dds 
- fix packaged tools can be executed by aligning the paths with the package structure (3.2.1-0002-fix-tool-paths.patch)
- add version ranges for tinyxml2 and asio 
- remove components to use only a main component as only one library is packaged (not sure if it makes sense to create components for the binaries)
- reworked the test package for 3.X as the API has significantly changed

The files in the recipes/fast-dds/3.x/test_package/msg directory have been generated based on the HelloWorld.idl within the same directory. This is done using a tool called `fastddsgen`. The old generated messages have been generated using an older version of `fastddsgen` which is not compatible with the Fast-DDS 3.X versions. This significant difference in the test_packages was the main reason to split fast-dds into 2.X and 3.X.

I hope that my explanation has helped in some way and that the review is now a little easier. Please let me know if anything is still unclear.

Thanks!

FannoFlow commented on recipes/fast-dds/3.x/conanfile.py:
```diff
@@ -0,0 +1,156 @@
+    @property
+    def _compilers_minimum_version(self):
+        return {
+            "gcc": "11",
+            "clang": "15",
+            "apple-clang": "15",
```
The required compiler versions should be relaxed. The only requirement is C++11 

https://github.com/conan-io/conan-center-index/pull/27887#discussion_r2194785573

nlhm commented:
Hi, I was wondering if there has been any further progress on adding version 3.2.x of Fast DDS? Thank you!

franramirez688 commented:
Superseded by https://github.com/conan-io/conan-center-index/pull/28124 🙏
